### DISCUSSION TITLE
Clean setup with Next11

### DISCUSSION THREAD

**Discussion by Duckinm** [payloadcms/payload#201] [Q&A]
How can I integrate with Next11? 

I already watched the boilerplates youtube videos and try them in many ways, but it seems not to work.

**Comment by bjorntheart** [comment]
This should get you started https://github.com/payloadcms/nextjs-custom-server

**Reply by Duckinm** [reply]
I tried, It's work now! Thank you.

By the way, Do you know how to make path alias work on this boilerplate?

**Reply by Ontopic** [reply]
Since the issue list is not that unmanageable I hope you guys can forgive me that I'm temporarily using this as a "draft" to once clear what the actual question is I can delete this and post an answer more inline with style guidelines and keeping things nicely organised, perhaps scavenge some of it for a doc entry or just a diary entry for myself 🙃

-----

To answer the question about path aliases, not sure if you mean import path aliases, path aliases in NextJS config (as in what they call rewrites and redirects) or the file-based path aliasing or catch-all-files in NextJS, or even more things that could be called "make path alias work" in a fair way.

So I'm just gonna share my thoughts on all of it since I feel worth sharing insights on all of them is worth it anyway, you just let me know what you're looking to achieve and I can delete this message and just post the actual answer / relevant advice instead and maybe rewrite the rest down in a structured way for the docs at some point

1. Import aliasing
I implemented both custom import aliases (server and client) through one babel.config.js that is based on an ENV flag loading a shared Babel config with a few specialised additions for either Payload or NextJS to make them close enough to working the exact same. This way I can apart from having the same things available (like a styled-components plugin for Babel) make also sure Payload and NextJS both have access to the same import paths (making it easier to share code from - for example - the collection definition from Payload and it's rendering function for the CMS in the NextJS frontend).
Another thing is that you can extend the tsconfig.json files from the nextjs-custom-server repo to add import aliases to compilerOptions. I use that for when I don't want Payload and NextJS to be thinking they can access the code from the CMS / api /backend layer, like in almost any other setup would be impossible to even think of, so important to at least not just share everything without carefully thinking of the implication and separating what really needs to be separated.
Normally I would say import aliasing is something to carefully think about twice anyway. Given the unique opportunity that Payload offers with NextJS to basically turn three or even four layers (backend, cms, api, frontend) into one repo is surprisingly strong though - especially with typing, but that still it feels wiser to devs working to just trust that I'll make sure they can find every import in a way that just follows simple rules like "server layer just use `import { toSlug } from 'server/collections/Post'` than to instead let them going on a hunt and trying making relative imports from the Payload layer work into NextJS themselves because the "function they want is already in the server folder, why would I copy past it for the frontend?" to me is more safely answered with, "you can just import it from the server/client/shared alias, silly, they actually are safer than relative imports, because every layer can only be accessed through the alias for importing stuff declared as safe to share by the 'code owner' and should catch any issues"

2. NextJS file system aliasing / catch-all
Quite simply should all just work the same, just follow the docs on NextJS and use `[...parts]`

3. I feel the first thing I would think of hearing "path alias" is...
...try Googling `ln -s`, but that is now just to let you know I really need to have a little bit more info to know what it is you're asking 🤗
---
Some unrequested words of advice; I can not stress enough to devs giving Payload a shot and going, “uhmm, where is the page where it basically says ‘do it like this with NextJS, here is how you swap our react with Vue / Preact’ or every other combo / issue I can think of?” - stick with it for a bit. Trust me, even though those examples are not here yet, you’ll see it’s just a matter of time for more docs and examples will show up. Although you now need to dive into the code to figure some stuff out for yourself sometimes for now, at least I'm not regretting in investing diving into their code and finding it's actually a very well thought out piece of work in which no details were rushed or overlooked. I would choose that any day over a CLI tool (that just hides what's really going on anyway) or so many boilerplates it's actually no longer a help, but a serious issue deciding what boilerplate to choose.

**Reply by Duckinm** [reply]
Thank you for your kind answers and also sorry for the misleading. It's my mistake.

Honestly, My English isn’t that good. I already read What you're sharing and no.1 are close to me I guess. but I rather say it seems no to be on my point. (In case I understand your messages correctly)

I want this to be work on my project.
[(https://nextjs.org/docs/advanced-features/module-path-aliases)]

I will describe the issue that I'm facing:

As a Frontend Developer that hasn't experienced using expressJS as a Typescript almost Javascript. I initiate the project by using the core packages from the boilerplates "nextjs-custom-server". 

Everything work pretty well. the only problem is When I set the baseUrl and paths on tsconfig.json as I normally do on NextJS and Typescript project. Obviously, the console can't catch it. (Can't find modules).

so, I quickly handle it by googling... I using "module-alias" set it up on package.json as it docs do. Okay, now it works! until then Webpack shows me other errors and I don't know how to fix them.

![Screen Shot 2564-06-20 at 4 24 30 PM](https://user-images.githubusercontent.com/29366271/122668930-cbfc7280-d1e4-11eb-95cb-35f8648c9498.png)

**Reply by Ontopic** [reply]
So yes, answer 1 is what we're looking for :)

So the tsconfig.json you're using is not meant to be used in really setting up aliases to access dependencies from the Payload folder, or at least I would have to dive into the example repo to check again how exactly it is setup again when you start fresh. I'll do some better research, for me at least using a shared babel.config.js is more what felt like a solution, but this might help you until I freshen my answer up

tsconfig.js
```{
  "compilerOptions": {
    "baseUrl": ".",
    "paths": {
      "x": [
        "x.js"
      ],
      "@client/*": [
        "client/*"
      ],
      "@server/*": [
        "server/*"
      ],

      "@/client/*": [
        "client/*"
      ],

      "@slate-plugins": [
        "../packages/slate-plugins/*"        
      ]
    },
    ...
  }

**Reply by Duckinm** [reply]
After digging many hours to find something that works and also shouldn't be more complex as it should.

Finally, I found the optimal solution for this "absolute import" issue. Though, it doesn't feel complete for me, Because there is some mirroring between Webpack and tsconfig. This could be better in the future. But for now, I'm very fine.

We need one more dependency which is [tsconfig-paths](https://www.npmjs.com/package/tsconfig-paths)

and edit the "dev" script on package.json from "node dev.js" to "node -r tsconfig-paths/register dev.js"

Then adding something like this.

![Screen Shot 2564-06-21 at 4 17 50 AM](https://user-images.githubusercontent.com/29366271/122688656-bec49f80-d247-11eb-86ec-026c5370fdb9.png)

To make it work like a charm, You've to add an alias on payload.config.ts as shown in the picture above. also you need to make it similarly in tsconfig.json too

Examples of absolute path in tsconfig.json or jsconfig.json .
[next-absolute-path](https://nextjs.org/docs/advanced-features/module-path-aliases)

You're all set!

**Reply by jmikrut** [reply]
Hi all,

I absolutely love the depth of this conversation. As someone who has tried to do these same things with Payload, I think that the conversation has built up to reach the same conclusion that my team and I have come to independently. 

The conclusion is....to set aliases up properly in a TypeScript / Webpack project, you need to set up the same aliases in _**at least**_ two places:

1. webpack aliases, as @Duckinm has shown above
2. tsconfig paths, as @Ontopic has shown above

If you use ESLint, you should also set them up there as well. So that's possibly three locations that all need aliases set up in exactly the same way. Yikes. That's the world we live in, at least for now! Payload can do little in the way of optimizing that, unfortunately. 

Also - @Ontopic:

> Trust me, even though those examples are not here yet, you’ll see it’s just a matter of time for more docs and examples will show up

You are 100% correct. Many more examples and guides are on their way. We are actively planning to _significantly_ ramp up efforts like this as I write this. On behalf of everyone at Payload, we REALLY appreciate your interaction and your support!

**Reply by Ontopic** [reply]
As said, tsconfig.json is for me just an "extra", merging NextJS and Payload's babel.config.js is what I use to get shared module aliases.

It goes something like this (contrived example, definitely worth a revisit when the babel.config.js pull is released for a cleaner example).

babel.config.js
```
const nextjsPreset = require('next/babel')

/*
TODO 😇
const payloadPreset = require('payload/babel')
*/

const moduleResolverPlugin = [
  require.resolve('babel-plugin-module-resolver'),
  {
    root: ['./'], // will make import from 'server/...' or 'client/...' workd
    alias: {
      server: './server/*',

      // pro-tip: add directory-named-webpack-plugin in next.config.js and payload.config.js to webpack step for super human imports without having a million index.js files
      library: './library/library.js',
      hooks: './hooks/hooks.js',
       ...
   },
   extensions: ['.js', '.jsx', '.tsx'],
]

const extraPlugins = [
  require.resolve('@babel/plugin-proposal-nullish-coalescing-operator'),
  require.resolve('@babel/plugin-proposal-do-expressions'),
  require.resolve('babel-plugin-inline-react-svg'),

  [require.resolve('babel-plugin-styled-components'),  {
     ssr: true,
     displayName: true,
     preprocess: false
  }],
]
...

const payloadConfig = {
  presets: [
    require.resolve('@babel/preset-typescript'),
    [
      ...copyPaste from payload/babel.config.js, //  turn into require when ready
    ]
  ],
  plugins: [
    ...copyPaste from payload/babel.config.js, // turn into require when ready

    ...extraPlugins,

    moduleResolverPlugin,
  ],
}

...

module.exports = api => {
   api.cache(true)
   

   // below is just meant for my setup, where my devops guys requested still being able to tear apart payload from the same container as nextjs, in case it turns out that local reads (which I love too much for dev) in production act up, so we can split things without breaking
   if (process.env['SERVICE_TO_START'] === 'server')
      ...
   else if   if (process.env['SERVICE_TO_START'] === 'client)
      ...

   const nextjsConfig = nextjsPreset.default(api)

   const config = {
      ...nextjsConfig,
      plugins: [
        ...nextjsConfig?.plugins,

        ...extraPlugins,

        moduleResolverPlugin,
      ],
    }

    ...

    return config
  }
}

```

**Reply by Ontopic** [reply]
Just not sure about eslint, I pretend that I don't need it 👻 , apart from that this works without duplicating configs anywhere, although type annotations may also not be as good in certain IDE's if you use Typescript for project level files and alias them in babel.config.js (I personally love Typescript for libraries / open source stuff, but for actual implementation for clientsI think it is often not worth the investment)

I started with the Babel setup because I also really like my do expressions and styled components, organising module aliases then for me felt like a "two for the price of one" solution.

@Duckinm definitely gonna play around with your solution as well, nice you figured it out! I feel with a combo of all the different solutions there's enough possibilities to cure the hatred of `../../../.. ....` in anybody's preferred flavour

**Reply by henrituan** [reply]
Hi @Duckinm,  thanks for your code snippets. I want to ask if you have any trouble building the app for production ?

For me, absolute imports  work perfectly on dev (`nodemon`), but not with build commands such as `payload build` or ` payload generate:types`.

For more info: I do have `tsconfig-paths` in my `nodemon.json`:
```
{
  "ext": "ts",
  "exec": "node -r tsconfig-paths/register -r ts-node/register src/server.ts"
}
````

However, I don't know how to configure `tsconfig-paths` for `payload build` command. I've tried this one but no luck:
`cross-env PAYLOAD_CONFIG_PATH=src/payload.config.ts payload -r ts-node/register -r tsconfig-paths/register build`.



Do you have any idea what did I miss ?

Thanks a lots

**Reply by Duckinm** [reply]
I successfully make it work in production without any issues. Honestly, the way I setup was too far from my brain memory! 

Btw, you can see my old repository to figure this out yourself.
https://github.com/Duckinm/payload-next/
